src/dia_model.cpp

```diff
@@ -0,0 +1,890 @@
+        } else {
+            out = ggml_concat(ctx, out, ggml_mul_mat(ctx, model->decoder->heads[i], cur), 2);
+        }
+    }
+    struct ggml_tensor * cond = ggml_cont(ctx, ggml_view_2d(ctx, out, out->ne[0], out->ne[2], out->nb[2], 0));
+    struct ggml_tensor * uncond = ggml_cont(ctx, ggml_view_2d(ctx, out, out->ne[0], out->ne[2], out->nb[2], out->nb[1]));
+    return ggml_map_custom2(ctx, cond, uncond, &cfg_scale, out->ne[0], &model->cfg_scale);
+}
+
+static struct ggml_tensor * build_dia_encoder(ggml_context * ctx, dia_model * model, dia_context * dctx, dia_ubatch & batch) {
+    dctx->inp_tokens = ggml_new_tensor_1d(ctx, GGML_TYPE_I32, model->max_encoder_context_length*2);
+    ggml_set_input(dctx->inp_tokens);
+
+    dctx->encode_positions = ggml_new_tensor_1d(ctx, GGML_TYPE_I32, model->max_encoder_context_length);
+    ggml_set_input(dctx->encode_positions);
+
+    struct ggml_tensor * attn_mask = build_dia_encoder_attn_mask(ctx, dctx, model);
+
+    struct ggml_tensor * cur = ggml_reshape_3d(ctx, ggml_get_rows(ctx, model->encoder->embedding, dctx->inp_tokens), model->encoder_hidden_size, model->max_encoder_context_length, 2);
+    for (auto layer : model->encoder->layers) {
+        struct ggml_tensor * residual = cur;
+        
+        cur = dia_layer_norm(ctx, cur, layer->self_attn_norm);
+        // self-attention
+        {
+            struct ggml_tensor * Qcur = ggml_mul_mat(ctx, layer->q, cur);
+            struct ggml_tensor * Kcur = ggml_mul_mat(ctx, layer->k, cur);
+            struct ggml_tensor * Vcur = ggml_mul_mat(ctx, layer->v, cur);
+
+            // Strangely Dia follows the neoX Rotary Positional Embeddings Protocol
+            Qcur = ggml_rope(ctx, ggml_cont(ctx, ggml_reshape_4d(ctx, Qcur, model->head_size, model->encoder_attn_heads, model->max_encoder_context_length, 2)), dctx->encode_positions, model->head_size, 2);
+            Kcur = ggml_rope(ctx, ggml_cont(ctx, ggml_reshape_4d(ctx, Kcur, model->head_size, model->encoder_attn_heads, model->max_encoder_context_length, 2)), dctx->encode_positions, model->head_size, 2);
+            struct ggml_tensor * q = ggml_cont(ctx, ggml_permute(ctx, Qcur, 0, 2, 1, 3));
+            struct ggml_tensor * k = ggml_cont(ctx, ggml_permute(ctx, Kcur, 0, 2, 1, 3));
+            struct ggml_tensor * kq = ggml_mul_mat(ctx, k, q);
+            kq = ggml_soft_max_ext(ctx, kq, attn_mask, 1.0f, 0.0f);
+            struct ggml_tensor * v = ggml_cont_4d(ctx, ggml_transpose(ctx, Vcur), model->max_encoder_context_length, model->head_size, model->encoder_attn_heads, 2);
+            struct ggml_tensor * kqv = ggml_mul_mat(ctx, kq, v);
+            struct ggml_tensor * kqv_merged = ggml_permute(ctx, kqv, 2, 0, 1, 3);
+
+            // It is unclear why the attention ops in Dia's encoder don't project to the embedding dimension size as is standard. Instead they up project to the decoder's embedding dimension
+            // then down project back the the encoder embedding dimension. 
```

Comment:
If I had to guess, their dropout was too high during initial training because it really shouldn't be necessary to do this. This does imply that quantization should work well with the model.